python/pycrdt/_awareness.py

```diff
@@ -0,0 +1,182 @@
+
+        self._subscriptions = []
+
+    @property
+    def states(self) -> dict[int, dict[str, Any]]:
+        return self._states
+
+    @property
+    def user(self) -> dict[str, str] | None:
+        return self._user
+
+    @user.setter
+    def user(self, user: dict[str, str]):
+        self._user = user
+        self.set_local_state_field("user", self._user)
+
+    def get_changes(self, message: bytes) -> dict[str, Any]:
+        """
+        Update the states with a user state.
+        This function send the changes to subscribers.
+
+        Args:
+            msg: Bytes representing the user state.
+        """
+        message = read_message(message)
+        decoder = Decoder(message)
+        timestamp = int(time.time() * 1000)
+        added = []
+        updated = []
+        filtered_updated = []
+        removed = []
+        states = []
+        length = decoder.read_var_uint()
+        for _ in range(length):
+            client_id = decoder.read_var_uint()
+            clock = decoder.read_var_uint()
+            state_str = decoder.read_var_string()
+            state = None if not state_str else json.loads(state_str)
+            if state is not None:
+                states.append(state)
+            client_meta = self.meta.get(client_id)
+            prev_state = self._states.get(client_id)
+            curr_clock = 0 if client_meta is None else client_meta["clock"]
+            if curr_clock < clock or (
+                curr_clock == clock and state is None and client_id in self._states
+            ):
+                if state is None:
+                    if client_id == self.client_id and self._states.get(client_id) is not None:
+                        clock += 1
+                    else:
+                        if client_id in self._states:
+                            del self._states[client_id]
+                else:
+                    self._states[client_id] = state
+                self.meta[client_id] = {
+                    "clock": clock,
+                    "last_updated": timestamp,
+                }
+                if client_meta is None and state is not None:
+                    added.append(client_id)
+                elif client_meta is not None and state is None:
+                    removed.append(client_id)
+                elif state is not None:
+                    if state != prev_state:
+                        filtered_updated.append(client_id)
+                    updated.append(client_id)
+
+        changes = {
+            "added": added,
+            "updated": updated,
+            "filtered_updated": filtered_updated,
+            "removed": removed,
+            "states": states,
+        }
+
+        # Do not trigger the callbacks if it is only a keep alive update
+        if added or filtered_updated or removed:
+            for callback in self._subscriptions:
+                callback(changes)
+
+        return changes
+
+    def get_local_state(self) -> dict[str, Any]:
```

Comment:
There should be a docstring.

Reply:
Done in https://github.com/jupyter-server/pycrdt/pull/170/commits/d20e98b7f2f837292a1e373ba5c00665d2dc02b7